core/schema/service.go

```diff
@@ -5,84 +5,100 @@ import (
 	"runtime/debug"
 
 	"github.com/dagger/dagger/core"
+	"github.com/dagger/dagger/dagql"
 )
 
 type serviceSchema struct {
-	*APIServer
-
-	svcs *core.Services
+	srv *dagql.Server
 }
 
 var _ SchemaResolvers = &serviceSchema{}
 
-func (s *serviceSchema) Name() string {
-	return "service"
+func (s *serviceSchema) Install() {
+	dagql.Fields[*core.Container]{
+		dagql.Func("asService", s.containerAsService).
+			Doc(`Turn the container into a Service.`,
+				`Be sure to set any exposed ports before this conversion.`),
+	}.Install(s.srv)
+
+	dagql.Fields[*core.Service]{
+		dagql.NodeFunc("hostname", s.hostname).
+			Doc(`Retrieves a hostname which can be used by clients to reach this container.`),
+
+		dagql.NodeFunc("ports", s.ports).
```

Comment:
If the service is a host tunnel and frontend port is 0 so that a random port on the host is chosen, then calling `start`/`stop` repeatedly (or separately in parallel for whatever reason someone would do that), then conceivably the ports value could be different each time right?

Does the fact that `start`/`stop` are impure mean that all subsequent calls chained after are also impure? If so, then I guess this is already handled probably. Otherwise I think it would make sense to mark `ports`+`endpoint` as impure too.

Pretty obscure obviously, mostly asking for my own understanding.

Reply:
Haha these are fun examples.

If `Start`/`Stop` supported chaining in the .Start/.Stop sense the impurity would carry over, but they return scalars, so they're really just for their side effect. Either way it's true that if you were doing that for whatever reason (naive/abstract automation) you definitely wouldn't want `ports` or `endpoint` to be cached.

So I think yeah, these should just be `Impure()`. Keeping in mind that `Impure()` is basically how everything works today so it's always a safe bet.

Reply:
also poking @jedevc for his own vindication since this is an example where for tunnels we'd want different behavior from regular services, in this case with regard to measuring purity.